Mflow

cschroed-usgs opened the pull request:
A lot of this diff is due to a rename. I will point out the relevant bits with line comments

cschroed-usgs commented on nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/dao/MflowDao.java:
```diff
@@ -31,4 +36,23 @@
+	 * @return null if unavailable, otherwise return a fully-formed 
+	 * WaterYearInterval
+	 */
+	public WaterYearAndMonthInterval getAvailability(String siteQwId){
```
The service definition

cschroed-usgs replied:
*DAO. The DAO definition

cschroed-usgs commented on nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/WaterYearAndMonthInterval.java:
```diff
@@ -0,0 +1,42 @@
+package gov.usgs.cida.nar.mybatis.model;
+
+
+public class WaterYearAndMonthInterval extends WaterYearInterval {
```
The obj for mybatis mapping

cschroed-usgs commented on nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/WaterYearInterval.java:
```diff
@@ -6,39 +6,39 @@
  */
 public class WaterYearInterval extends NARData {
 
```
Renamed this from start->startYear, end->endYear so that start and end months could unambiguously extend this class.

cschroed-usgs commented on nar-webservices/src/main/java/gov/usgs/cida/nar/service/MflowService.java:
```diff
@@ -66,10 +70,32 @@ public TimeStepDensity getTimeStepDensity() {
 	public TimeSeriesCategory getTimeSeriesCategory() {
```
Here's the real service definition

cschroed-usgs commented on nar-webservices/src/main/java/gov/usgs/cida/nar/service/MflowService.java:
```diff
@@ -66,10 +70,32 @@ public TimeStepDensity getTimeStepDensity() {
 		return TimeSeriesCategory.FLOW;
 	}
-
+	private boolean allTimeIntervalFieldsInitialized(WaterYearAndMonthInterval interval) {
```
Is this method well name? Does it belong here as a method in the service layer, or as a static or instance method of WaterYearAndMonthInterval?

jiwalker-usgs replied:
This probably makes more sense in the WaterYearAndMonthInterval class and naming it something like isInititialized or something.

cschroed-usgs commented on nar-webservices/src/main/java/gov/usgs/cida/nar/service/MflowService.java:
```diff
@@ -66,10 +70,32 @@ public TimeStepDensity getTimeStepDensity() {
+				this.getTimeStepDensity(),
+				startTime,
+				endTime,
+				null
```
reminder: null == no constituent

cschroed-usgs commented on nar-webservices/src/main/resources/gov/usgs/cida/nar/mybatis/mappers/MflowMapper.xml:
```diff
@@ -35,4 +42,30 @@
 	ORDER BY site_qw_id, site_flow_id, wy, month ASC
 	</select>
 	
+	<select id="getAvailability" parameterType="map" resultMap="waterYearAndMonthInterval">
```
This query is kinda gnarly. I can simplify this to a simple `SELECT min(date), max(date)...` query by modifying the serialization function in nar_data so that it takes YEAR and MONTH fields and writes them out in a single ISO-compliant string. Liquibase can then parse it, put it into a postgres DATETIME column, and the query can turn dead-simple. Should I do that?

jiwalker-usgs replied:
I've avoided playing with the data much and just mapping the format as-is to the database.  I think there is a good case to be made for doing so, but I think actually converting to a date type might result in different problems.  The other trick here is that I'm pretty sure month is always 5 so much of this just being defensive.

jiwalker-usgs commented:
Looks good.  One thing I'm thinking while looking at this is whether it would make sense to change the DataAvailability object from one that maps to the database to one that is closer to the service layer.  We could then do some shared queries (i.e. max wy) and compose the results into something at the service layer.  Just a thought.

cschroed-usgs commented on nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/WaterYearAndMonthInterval.java:
```diff
@@ -0,0 +1,50 @@
+	public void setEndMonth(Integer endMonth) {
+		this.endMonth = endMonth;
+	}
+	
```
@jiwalker-usgs , I moved it here to the instance.

cschroed-usgs commented on nar-webservices/src/test/java/gov/usgs/cida/nar/mybatis/model/WaterYearAndMonthIntervalTest.java:
```diff
@@ -0,0 +1,36 @@
+		beanTester.testBean(TimeSeriesAvailability.class);
+	}
+	
+	@Test
```
I also added a quick test here

cschroed-usgs commented:
Thanks for the feedback here and for chatting in person. I moved the method and added tests. I will contact the PL about pushing back on the month column. I will react to the other feedback you gave based on their response.
